File: rh-containers-x-container-2/runner/search_research.js
const path = require('path');
const {
  launchContext,
  writeJson,
  getBody,
  looksLoggedIn,
  collectVisiblePosts,
  jitter,
  OUT_DIR,
} = require('./lib');

const QUERY = process.env.X_QUERY || '';
const LIMIT = Number(process.env.X_LIMIT || '10');

(async () => {
  const result = { status: 'unknown', query: QUERY, limit: LIMIT };
  let browser;
  let context;
  try {
    ({ browser, context } = await launchContext());
    const page = context.pages()[0] || await context.newPage();
    await page.goto('https://x.com/home', { waitUntil: 'domcontentloaded', timeout: 45000 });
    await page.waitForTimeout(jitter(4000, 3000));
    const homeBody = await getBody(page);
    if (!looksLoggedIn(homeBody)) {
      result.status = 'fail';
      result.reason = 'session-not-authenticated';
    } else if (!QUERY) {
      result.status = 'fail';
      result.reason = 'missing-query';
    } else {
      const searchUrl = `https://x.com/search?q=${encodeURIComponent(QUERY)}&src=typed_query&f=live`;
      await page.goto(searchUrl, { waitUntil: 'domcontentloaded', timeout: 45000 });
      await page.waitForTimeout(jitter(6000, 5000));
      const body = await getBody(page);
      const posts = await collectVisiblePosts(page, LIMIT);
      result.status = 'pass';
      result.reason = 'search-collected';
      result.url = page.url();
      result.title = await page.title();
      result.excerpt = body.slice(0, 1600);
      result.posts = posts;
      await page.screenshot({ path: path.join(OUT_DIR, 'search_research.png'), fullPage: true });
    }
  } catch (error) {
    result.status = 'error';
    result.reason = String(error);
  }
  writeJson('search_research.json', result);
  console.log(JSON.stringify(result, null, 2));
  if (browser) await browser.close().catch(() => {});
})();

File: rh-containers-x-container-1/runner/search_research.py
from __future__ import annotations

import asyncio
import json
import os
from urllib.parse import quote

from mcp_browser import ChromeMCPBrowser, jitter, looks_logged_in, scroll_and_collect
from shared import OUT_DIR, resolve_browser_url, write_json


QUERY = os.environ.get("X_QUERY", "").strip()
LIMIT = int(os.environ.get("X_LIMIT", "10"))


async def main() -> None:
    result = {"status": "unknown", "query": QUERY, "limit": LIMIT}
    browser_url = resolve_browser_url()

    try:
        async with ChromeMCPBrowser(browser_url) as browser:
            await browser.navigate("https://x.com/home")
            await browser.sleep(jitter(4000, 3000))
            home = await browser.get_page_payload(2000)
            if not looks_logged_in(home.get("text", "")):
                result["status"] = "fail"
                result["reason"] = "session-not-authenticated"
            elif not QUERY:
                result["status"] = "fail"
                result["reason"] = "missing-query"
            else:
                search_url = f"https://x.com/search?q={quote(QUERY)}&src=typed_query&f=live"
                await browser.navigate(search_url)
                await browser.sleep(jitter(6000, 5000))
                page = await browser.get_page_payload(1600)
                posts = await scroll_and_collect(
                    browser,
                    seen_tweet_ids=set(),
                    max_scrolls=8,
                    posts_per_scroll=max(LIMIT, 20),
                )
                result["status"] = "pass"
                result["reason"] = "search-collected"
                result["url"] = page.get("url")
                result["title"] = page.get("title")
                result["excerpt"] = page.get("text", "")[:1600]
                result["posts"] = posts[:LIMIT]
                screenshot_path = OUT_DIR / "search_research.png"
                await browser.screenshot(str(screenshot_path), full_page=True)
    except Exception as error:
        result["status"] = "error"
        result["reason"] = str(error)

    write_json("search_research.json", result)
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    asyncio.run(main())
